src/metabase/util/ssh.clj

```diff
@@ -1,32 +1,63 @@
-                                                               ^String tunnel-user
-                                                               ^String tunnel-host
-                                                               tunnel-port)
-                     (.setPassword ^String tunnel-pass)
-                     (.setConfig "StrictHostKeyChecking" "no")
-                     (.connect default-ssh-timeout)
-                     (.setPortForwardingL 0 host port))
-        input-port (some-> (.getPortForwardingL connection)
-                           first
-                           (str/split #":")
-                           first
-                           (Integer/parseInt))]
-    (assert (number? input-port))
-    (log/info (u/format-color 'cyan "creating ssh tunnel %s@%s:%s -L %s:%s:%s" tunnel-user tunnel-host tunnel-port input-port host port))
-    [connection input-port]))
+  [{:keys [tunnel-host tunnel-port tunnel-user tunnel-pass tunnel-private-key tunnel-private-key-passphrase host port]}]
+  (let [conn-future (.connect ^SshClient client ^String tunnel-user ^String tunnel-host ^Integer tunnel-port)
+        conn-status (.verify ^ConnectFuture conn-future default-ssh-timeout)
+        session (.getSession ^SessionHolder conn-status)
+
+        _ (when tunnel-pass
+            (.addPasswordIdentity ^ClientSession session tunnel-pass))
+        _ (when tunnel-private-key
+            (let [resource-key (proxy [AbstractIoResource] [(class "key") "key"])
+                  password-provider (proxy [FilePasswordProvider] []
+                                      (getPassword [_ _ _]
+                                        tunnel-private-key-passphrase)
+                                      (handleDecodeAttemptResult [_ _ _ _ _]
+                                        FilePasswordProvider$ResourceDecodeResult/TERMINATE))
+
+                  ids (SecurityUtils/loadKeyPairIdentities ^ClientSession session
+                                                           resource-key
+                                                           (ByteArrayInputStream. (.getBytes ^String tunnel-private-key "UTF-8"))
+                                                           password-provider)
+
+                  keypair (GenericUtils/head ids)]
+              (.addPublicKeyIdentity ^ClientSession session keypair)))
+
+        auth-future (.auth ^ClientSession session)
+        _ (.verify ^AuthFuture auth-future default-ssh-timeout)
+
+        tracker (.createLocalPortForwardingTracker ^ClientSession session
+                                                   (SshdSocketAddress. "" 0)
+                                                   (SshdSocketAddress. host port))
+
+        input-port (.getPort ^SshdSocketAddress (.getBoundAddress ^PortForwardingTracker tracker))]
+    (log/trace (u/format-color 'cyan "creating ssh tunnel %s@%s:%s -L %s:%s:%s" tunnel-user tunnel-host tunnel-port input-port host port))
+    [session tracker]))
```

Comment:
Just my two cents: I think this whole function would be a lot more readable if you broke some of the stuff out into separate functions and used `doto` liberally.

```clj
(def ^:private ^SshClient client
  (doto (SshClient/setUpDefaultClient)
    (.start)
    (.setForwardingFilter AcceptAllForwardingFilter/INSTANCE)))

(defn- maybe-add-tunnel-password! [^ClientSession session {:keys [^String tunnel-pass]}]
  (when tunnel-pass
    (.addPasswordIdentity session tunnel-pass)))

(defn- maybe-add-tunnel-private-key! [^ClientSession session ^String tunnel-private-key tunnel-private-key-passphrase]
  (when tunnel-private-key
    (let [resource-key      (proxy [AbstractIoResource] [(class "key") "key"])
          password-provider (proxy [FilePasswordProvider] []
                              (getPassword [_ _ _]
                                tunnel-private-key-passphrase)
                              (handleDecodeAttemptResult [_ _ _ _ _]
                                FilePasswordProvider$ResourceDecodeResult/TERMINATE))
          ids               (with-open [is (ByteArrayInputStream. (.getBytes tunnel-private-key "UTF-8"))]
                              (SecurityUtils/loadKeyPairIdentities session resource-key is password-provider))
          keypair           (GenericUtils/head ids)]
      (.addPublicKeyIdentity session keypair))))

(defn start-ssh-tunnel!
  "Opens a new ssh tunnel and returns the connection along with the dynamically assigned tunnel entrance port. It's the
  callers responsibility to call `close-tunnel` on the returned connection object."
  [{:keys [^String tunnel-host ^Integer tunnel-port ^String tunnel-user tunnel-pass tunnel-private-key
           tunnel-private-key-passphrase host port]}]
  (let [^ConnectFuture conn-future (.connect client tunnel-user tunnel-host tunnel-port)
        ^SessionHolder conn-status (.verify conn-future default-ssh-timeout)
        session                    (doto ^ClientSession (.getSession conn-status)
                                     (maybe-add-tunnel-password! tunnel-pass)
                                     (maybe-add-tunnel-private-key! tunnel-private-key tunnel-private-key-passphrase)
                                     (.. auth (verify default-ssh-timeout)))
        tracker                    (.createLocalPortForwardingTracker session
                                                                      (SshdSocketAddress. "" 0)
                                                                      (SshdSocketAddress. host port))
        input-port                 (.. tracker getBoundAddress getPort)]
    (log/trace (u/format-color 'cyan "creating ssh tunnel %s@%s:%s -L %s:%s:%s" tunnel-user tunnel-host tunnel-port input-port host port))
    [session tracker]))
```

that's my attempt at tweaking it a bit. I haven't checked that it works but it doesn't cause any reflection warnings at least.

Reply:
huh I hadn't seen the `..` form before. I like that.